src/language-html/print/tag.js

```diff
@@ -0,0 +1,385 @@
+  const parts = [
+    indent([
+      forceNotToBreakAttrContent ? " " : line,
+      join(line, printedAttributes),
+    ]),
+  ];
+
+  if (
+    /**
+     *     123<a
+     *       attr
+     *           ~
+     *       >456
+     */
+    (node.firstChild &&
+      needsToBorrowParentOpeningTagEndMarker(node.firstChild)) ||
+    /**
+     *     <span
+     *       >123<meta
+     *                ~
+     *     /></span>
+     */
+    (node.isSelfClosing &&
+      needsToBorrowLastChildClosingTagEndMarker(node.parent)) ||
+    forceNotToBreakAttrContent
+  ) {
+    parts.push(node.isSelfClosing ? " " : "");
+  } else {
+    parts.push(
+      options.bracketSameLine
+        ? node.isSelfClosing
+          ? " "
+          : ""
+        : node.isSelfClosing
+        ? line
+        : softline
+    );
+  }
+
+  return parts;
+}
+
+function printOpeningTagEnd(node) {
+  return node.firstChild &&
+    needsToBorrowParentOpeningTagEndMarker(node.firstChild)
+    ? ""
+    : printOpeningTagEndMarker(node);
+}
+
+function printOpeningTag(path, options, print) {
+  const node = path.getValue();
+
+  return [
+    printOpeningTagStart(node, options),
+    printAttributes(path, options, print),
+    node.isSelfClosing ? "" : printOpeningTagEnd(node),
+  ];
+}
+
+function printOpeningTagStart(node, options) {
+  return node.prev && needsToBorrowNextOpeningTagStartMarker(node.prev)
+    ? ""
+    : [printOpeningTagPrefix(node, options), printOpeningTagStartMarker(node)];
+}
+
+function printOpeningTagPrefix(node, options) {
+  return needsToBorrowParentOpeningTagEndMarker(node)
+    ? printOpeningTagEndMarker(node.parent)
+    : needsToBorrowPrevClosingTagEndMarker(node)
+    ? printClosingTagEndMarker(node.prev, options)
+    : "";
+}
+
+function printOpeningTagStartMarker(node) {
+  switch (node.type) {
+    case "ieConditionalComment":
+    case "ieConditionalStartComment":
+      return `<!--[if ${node.condition}`;
+    case "ieConditionalEndComment":
+      return "<!--<!";
+    case "interpolation":
+      return "{{";
+    case "docType":
+      return "<!DOCTYPE";
+    case "element":
+      if (node.condition) {
+        return `<!--[if ${node.condition}]><!--><${node.rawName}`;
+      }
+    // fall through
+    default:
+      return `<${node.rawName}`;
+  }
+}
+
+function printOpeningTagEndMarker(node) {
+  assert(!node.isSelfClosing);
+  switch (node.type) {
+    case "ieConditionalComment":
+      return "]>";
+    case "element":
+      if (node.condition) {
+        return "><!--<![endif]-->";
+      }
+    // fall through
+    default:
+      return ">";
+  }
+}
+
+module.exports = {
+  closing: {
+    printClosingTag,
+    printClosingTagStart,
+    printClosingTagStartMarker,
+    printClosingTagEndMarker,
+    printClosingTagSuffix,
+    printClosingTagEnd,
+    needsToBorrowLastChildClosingTagEndMarker,
+    needsToBorrowParentClosingTagStartMarker,
+    needsToBorrowPrevClosingTagEndMarker,
+  },
+  opening: {
```

Comment:
I don't like `closing` and `opening`, when we move to esm, most  likely we'll use named export, why not put everything on `module.exports`?

Reply:
fixed 675228d